core/src/main/java/org/elasticsearch/client/transport/TransportClientNodesService.java

```diff
@@ -211,7 +211,13 @@ public TransportClientNodesService removeTransportAddress(TransportAddress trans
     }
 
     public <Response> void execute(NodeListenerCallback<Response> callback, ActionListener<Response> listener) {
-        List<DiscoveryNode> nodes = this.nodes;
+        List<DiscoveryNode> nodes;
+        synchronized (mutex) {
```

Comment:
I don't think it is worth synchronizing because the race condition is going to cause us to throw the wrong error, but only briefly and only if we change the order that we set the things? I do agree we can't rely on that order because it isn't documented and even if it were I'm liable to screw it up the next time I touch the code because I don't read good.

I wonder if we can make this less trappy and have a single volatile reference that we replace when we change state? There shouldn't be an ordering issue then, right? And a buffoon like me is less likely to screw it up the next time I touch it?